server/protocol/qdatastreamprotocol.py

```diff
@@ -127,31 +126,60 @@ def pack_message(*args: str) -> bytes:
 
     def close(self):
         """
-        Close writer stream
+        Close writer stream as soon as the buffer has emptied.
         :return:
         """
         self.writer.close()
 
+    def abort(self):
+        """
+        Close writer stream immediately discarding the buffer contents
+        :return:
+        """
+        self.writer.transport.abort()
+
+
+    async def drain(self):
+        """
+        Await the write buffer to empty.
+        See StreamWriter.drain()
+
+        :raises: DisconnectedError if the client disconnects while waiting for
+        the write buffer to empty.
+        """
+        # NOTE: This sleep is needed in python versions <= 3.6
+        # https://github.com/aio-libs/aioftp/issues/7
+        await asyncio.sleep(0)
+        async with self._drain_lock:
```

Comment:
When I see this two lines, I must say, that i'm not happy from `asyncio`... :disappointed: 
(or we're doing something wrong :see_no_evil: )

Reply:
It's an asyncio issue. `drain` doesn't support multiple waiters so we have to use a lock. This is actually the solution from the `websockets` library.